frontend/packages/dev-console/src/components/topology/topology-utils.ts

```diff
@@ -742,21 +742,26 @@ export const topologyModelFromDataModel = (
     };
   });
 
-  // create links from data
-  const edges = dataModel.graph.edges.map(
-    (d): EdgeModel => ({
-      data: d,
-      source: d.source,
-      target: d.target,
-      id: `${d.source}_${d.target}`,
-      type: d.type,
-    }),
-  );
+  // create links from data, only include those which have a valid source and target
+  const allNodes = [...nodes, ...groupNodes];
+  const edges = dataModel.graph.edges
+    .filter((d) => {
+      return allNodes.find((n) => n.id === d.source) && allNodes.find((n) => n.id === d.target);
+    })
```

Comment:
In what situation do we have edges without valid source or target?

Reply:
This occurred when we allowed (mistakenly) connections to knative revisions or for a node to itself. While those situations have been taken care of, this change safe guards against other potential similar situations that could occur.